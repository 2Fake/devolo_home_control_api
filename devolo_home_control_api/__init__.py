"""the devolo Home Control API."""
from importlib.metadata import PackageNotFoundError, version

try:
    __version__ = version("devolo_home_control_api")
except PackageNotFoundError:
    # package is not installed - e.g. pulled and run locally
    __version__ = "0.0.0"
