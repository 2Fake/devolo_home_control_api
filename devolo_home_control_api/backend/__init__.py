MESSAGE_TYPES = {
    "devolo.BinarySensor": "_binary_sensor",
    "devolo.BinarySwitch": "_binary_switch",
    "devolo.Blinds": "_multi_level_switch",
    "devolo.DevicesPage": "_inspect_devices",
    "devolo.DewpointSensor": "_multi_level_sensor",
    "devolo.Dimmer": "_multi_level_switch",
    "devolo.Grouping": "_grouping",
    "devolo.HumidityBarValue": "_humidity_bar",
    "devolo.HumidityBarZone": "_humidity_bar",
    "devolo.mprm.gw.GatewayAccessibilityFI": "_gateway_accessible",
    "devolo.Meter": "_meter",
    "devolo.MildewSensor": "_binary_sensor",
    "devolo.MultiLevelSensor": "_multi_level_sensor",
    "devolo.MultiLevelSwitch": "_multi_level_switch",
    "devolo.RemoteControl": "_remote_control",
    "devolo.SirenMultiLevelSwitch": "_multi_level_switch",
    "devolo.ShutterMovementFI": "_binary_sensor",
    "devolo.ValveTemperatureSensor": "_multi_level_sensor",
    "devolo.VoltageMultiLevelSensor": "_multi_level_sensor",
    "devolo.WarningBinaryFI": "_binary_sensor",
    "hdm": "_device_state",
    "acs.hdm": "_automatic_calibration",
    "bas.hdm": "_binary_async",
    "bss.hdm": "_binary_sync",
    "cps.hdm": "_parameter",
    "gds.hdm": "_general_device",
    "lis.hdm": "_led",
    "mas.hdm": "_multilevel_async",
    "mss.hdm": "_multilevel_sync",
    "ps.hdm": "_protection",
    "stmss.hdm": "_multilevel_sync",
    "sts.hdm": "_switch_type",
    "trs.hdm": "_temperature_report",
    "vfs.hdm": "_led"
}
