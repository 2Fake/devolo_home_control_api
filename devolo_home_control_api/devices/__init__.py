"""Devices appearing in a devolo Home Control setup."""
from .gateway import Gateway
from .zwave import Zwave

__all__ = ["Gateway", "Zwave"]
