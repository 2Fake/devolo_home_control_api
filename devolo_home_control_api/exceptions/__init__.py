"""Expections used in this package."""
from .device import SwitchingProtected, WrongElementError
from .gateway import GatewayOfflineError
from .general import WrongCredentialsError, WrongUrlError

__all__ = ["GatewayOfflineError", "SwitchingProtected", "WrongCredentialsError", "WrongElementError", "WrongUrlError"]
