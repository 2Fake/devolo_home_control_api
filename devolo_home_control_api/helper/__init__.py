"""Helper functions used in the package."""
from .names import camel_case_to_snake_case
from .uid import (
    get_device_type_from_element_uid,
    get_device_uid_from_element_uid,
    get_device_uid_from_setting_uid,
    get_home_id_from_device_uid,
    get_sub_device_uid_from_element_uid,
)

__all__ = [
    "camel_case_to_snake_case",
    "get_device_type_from_element_uid",
    "get_device_uid_from_element_uid",
    "get_device_uid_from_setting_uid",
    "get_home_id_from_device_uid",
    "get_sub_device_uid_from_element_uid",
]
