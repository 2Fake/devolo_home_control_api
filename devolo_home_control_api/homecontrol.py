import threading
from typing import Dict, Optional

import requests
from zeroconf import Zeroconf

from . import __version__
from .backend import MESSAGE_TYPES
from .backend.mprm import Mprm
from .devices.gateway import Gateway
from .devices.zwave import Zwave
from .helper.string import camel_case_to_snake_case
from .helper.uid import (get_device_type_from_element_uid,
                         get_device_uid_from_element_uid,
                         get_device_uid_from_setting_uid,
                         get_home_id_from_device_uid)
from .mydevolo import Mydevolo
from .properties.binary_sensor_property import BinarySensorProperty
from .properties.binary_switch_property import BinarySwitchProperty
from .properties.consumption_property import ConsumptionProperty
from .properties.humidity_bar_property import HumidityBarProperty
from .properties.multi_level_sensor_property import MultiLevelSensorProperty
from .properties.multi_level_switch_property import MultiLevelSwitchProperty
from .properties.remote_control_property import RemoteControlProperty
from .properties.settings_property import SettingsProperty
from .publisher.publisher import Publisher
from .publisher.updater import Updater


class HomeControl(Mprm):
    """
    Representing object for your Home Control setup. This is more or less the glue between your devolo Home Control Central
    Unit, your devices and their properties.

    :param gateway_id: Gateway ID (aka serial number), typically found on the label of the device
    :param mydevolo_instance: Mydevolo instance for talking to the devolo Cloud
    :param zeroconf_instance: Zeroconf instance to be potentially reused
    """

    def __init__(self, gateway_id: str, mydevolo_instance: Mydevolo, zeroconf_instance: Optional[Zeroconf] = None):
        self._mydevolo = mydevolo_instance
        self._session = requests.Session()
        self._session.headers.update({"User-Agent": f"devolo_home_control_api/{__version__}"})
        self._zeroconf = zeroconf_instance
        self.gateway = Gateway(gateway_id, mydevolo_instance)

        super().__init__()
        self._grouping()

        # Create the initial device dict
        self.devices: Dict = {}
        self._inspect_devices(self.get_all_devices())

        self.device_names = {
            f"{self.devices[device].settings_property['general_device_settings'].name}/\
              {self.devices[device].settings_property['general_device_settings'].zone}": self.devices[device].uid
            for device in self.devices
        }

        self.gateway.home_id = get_home_id_from_device_uid(next(iter(self.device_names.values())))

        self.publisher = Publisher(self.devices.keys())

        self.updater = Updater(devices=self.devices, gateway=self.gateway, publisher=self.publisher)
        self.updater.on_device_change = self.device_change

        threading.Thread(target=self.websocket_connect, name=f"{self.__class__.__name__}.websocket_connect").start()
        self.wait_for_websocket_establishment()

    @property
    def binary_sensor_devices(self) -> list:
        """ Get all binary sensor devices. """
        return [self.devices[uid] for uid in self.devices if hasattr(self.devices[uid], "binary_sensor_property")]

    @property
    def binary_switch_devices(self) -> list:
        """ Get all binary switch devices. """
        return [self.devices[uid] for uid in self.devices if hasattr(self.devices[uid], "binary_switch_property")]

    @property
    def blinds_devices(self) -> list:
        """ Get all blinds devices. """
        blinds_devices = []
        for device in self.multi_level_switch_devices:
            blinds_devices.extend([
                self.devices[device.uid]
                for multi_level_switch_property in device.multi_level_switch_property
                if multi_level_switch_property.startswith("devolo.Blinds")
            ])
        return blinds_devices

    @property
    def multi_level_sensor_devices(self) -> list:
        """ Get all multi level sensor devices. """
        return [self.devices[uid] for uid in self.devices if hasattr(self.devices[uid], "multi_level_sensor_property")]

    @property
    def multi_level_switch_devices(self) -> list:
        """ Get all multi level switch devices. This also includes blinds devices. """
        return [self.devices[uid] for uid in self.devices if hasattr(self.devices[uid], "multi_level_switch_property")]

    @property
    def remote_control_devices(self) -> list:
        """ Get all remote control devices. """
        return [self.devices[uid] for uid in self.devices if hasattr(self.devices[uid], "remote_control_property")]

    def device_change(self, device_uids: list):
        """
        React on new devices or removed devices. As the Z-Wave controller can only be in inclusion or exclusion mode, we
        assume, that you cannot add and remove devices at the same time. So if the number of devices increases, there is
        a new one and if the number decreases, a device was removed.

        :param device_uids: List of UIDs known by the backend
        """
        if len(device_uids) > len(self.devices):
            devices = [device for device in device_uids if device not in self.devices]
            mode = "add"
            self._inspect_devices([devices[0]])
            self._logger.debug("Device %s added.", devices[0])
        else:
            devices = [device for device in self.devices if device not in device_uids]
            mode = "del"
            self.devices.pop(devices[0])
            self._logger.debug("Device %s removed.", devices[0])
        self.updater.devices = self.devices
        return (devices[0], mode)

    def on_update(self, message):
        """
        Initialize steps needed to update properties on a new message.

        :param message: Message because of which we need to update properties
        """
        self.updater.update(message)

    def _binary_sensor(self, uid_info: dict):
        """ Process BinarySensor properties"""
        device_uid = get_device_uid_from_element_uid(uid_info['UID'])
        if not hasattr(self.devices[device_uid], "binary_sensor_property"):
            self.devices[device_uid].binary_sensor_property = {}
        self._logger.debug("Adding binary sensor property to %s.", device_uid)
        self.devices[device_uid].binary_sensor_property[uid_info['UID']] = BinarySensorProperty(
            element_uid=uid_info['UID'],
            state=bool(uid_info['properties']['state']),
            sensor_type=uid_info['properties']['sensorType'],
            sub_type=uid_info['properties']['subType'])

    def _binary_switch(self, uid_info: dict):
        """ Process BinarySwitch properties. """
        device_uid = get_device_uid_from_element_uid(uid_info['UID'])
        if not hasattr(self.devices[device_uid], "binary_switch_property"):
            self.devices[device_uid].binary_switch_property = {}
        self._logger.debug("Adding binary switch property to %s.", device_uid)
        self.devices[device_uid].binary_switch_property[uid_info['UID']] = BinarySwitchProperty(
            element_uid=uid_info['UID'],
            setter=self.set_binary_switch,
            state=bool(uid_info['properties']['state']),
            enabled=uid_info['properties']['guiEnabled'])

    def _general_device(self, uid_info: dict):
        """ Process general device setting (gds) properties. """
        device_uid = get_device_uid_from_setting_uid(uid_info['UID'])
        self._logger.debug("Adding general device settings to %s.", device_uid)
        self.devices[device_uid].settings_property['general_device_settings'] = SettingsProperty(
            element_uid=uid_info['UID'],
            setter=self.set_setting,
            events_enabled=uid_info['properties']['settings']['eventsEnabled'],
            name=uid_info['properties']['settings']['name'],
            zone_id=uid_info['properties']['settings']['zoneID'],
            icon=uid_info['properties']['settings']['icon'],
            zones=self.gateway.zones)

    def _grouping(self):
        """ Get all zones (also called rooms). """
        self.gateway.zones = self.get_all_zones()

    def _humidity_bar(self, uid_info: dict):
        """
        Process HumidityBarZone and HumidityBarValue properties.

        For whatever reason, the zone and the position within that zone are two different FIs. Nevertheless, it does not make
        sense to separate them. So we fake an FI and a sensorType to combine them together into one object.
        """
        device_uid = get_device_uid_from_element_uid(uid_info['UID'])
        fake_element_uid = f"devolo.HumidityBar:{uid_info['UID'].split(':', 1)[1]}"
        if not hasattr(self.devices[device_uid], "humidity_bar_property"):
            self.devices[device_uid].humidity_bar_property = {}
        if self.devices[device_uid].humidity_bar_property.get(fake_element_uid) is None:
            self.devices[device_uid].humidity_bar_property[fake_element_uid] = HumidityBarProperty(
                element_uid=fake_element_uid,
                sensorType="humidityBar")
        if uid_info['properties']['sensorType'] == "humidityBarZone":
            self._logger.debug("Adding humidity bar zone property to %s.", device_uid)
            self.devices[device_uid].humidity_bar_property[fake_element_uid].zone = uid_info['properties']['value']
        elif uid_info['properties']['sensorType'] == "humidityBarPos":
            self._logger.debug("Adding humidity bar position property to %s.", device_uid)
            self.devices[device_uid].humidity_bar_property[fake_element_uid].value = uid_info['properties']['value']

    def _inspect_devices(self, devices: list):
        """ Inspect device properties of given list of devices. """
        devices_properties = self.get_data_from_uid_list(devices)
        for device_properties in devices_properties:
            properties = device_properties['properties']
            self.devices[device_properties['UID']] = Zwave(mydevolo_instance=self._mydevolo, **properties)
            self.devices[device_properties['UID']].settings_property = {}
            threading.Thread(target=self.devices[device_properties['UID']].get_zwave_info,
                             name=f"{self.__class__.__name__}.{self.devices[device_properties['UID']].uid}").start()

        # List comprehension gets the list of uids from every device
        nested_uids_lists = [(uid['properties'].get("settingUIDs") + uid['properties']['elementUIDs'])
                             for uid in devices_properties]

        # List comprehension gets all uids into one list to make one big call against the mPRM
        uid_list = [uid for sublist in nested_uids_lists for uid in sublist]

        device_properties_list = self.get_data_from_uid_list(uid_list)

        for uid_info in device_properties_list:
            message_type = MESSAGE_TYPES.get(get_device_type_from_element_uid(uid_info['UID']), "_unknown")
            getattr(self, message_type)(uid_info)
            try:
                uid = self.devices[get_device_uid_from_element_uid(uid_info['UID'])]
            except KeyError:
                uid = self.devices[get_device_uid_from_setting_uid(uid_info['UID'])]
            uid.pending_operations = uid.pending_operations or bool(uid_info['properties'].get("pendingOperations"))

        # Last activity messages sometimes arrive before a device was initialized and therefore need to be handled afterwards.
        [  # pylint: disable=expression-not-assigned
            self._last_activity(uid_info)
            for uid_info in device_properties_list
            if uid_info['UID'].startswith("devolo.LastActivity")
        ]

    def _automatic_calibration(self, uid_info: dict):
        """ Process automatic calibration (acs) properties. """
        device_uid = get_device_uid_from_setting_uid(uid_info['UID'])
        self._logger.debug("Adding automatic calibration setting to %s.", device_uid)
        self.devices[device_uid].settings_property['automatic_calibration'] = SettingsProperty(
            element_uid=uid_info['UID'],
            setter=self.set_setting,
            calibration_status=bool(uid_info['properties']['calibrationStatus']))

    def _binary_sync(self, uid_info: dict):
        """
        Process binary sync setting (bss) properties. Currently only the direction of a shutter in known to be a binary sync
        setting property, so it is named nicely.
        """
        device_uid = get_device_uid_from_setting_uid(uid_info['UID'])
        self._logger.debug("Adding binary sync setting to %s.", device_uid)
        self.devices[device_uid].settings_property['movement_direction'] = SettingsProperty(
            element_uid=uid_info['UID'],
            setter=self.set_setting,
            inverted=uid_info['properties']['value'])

    def _binary_async(self, uid_info: dict):
        """ Process binary async setting (bas) properties. """
        device_uid = get_device_uid_from_setting_uid(uid_info['UID'])
        self._logger.debug("Adding binary async settings to %s.", device_uid)
        settings_property = SettingsProperty(element_uid=uid_info['UID'],
                                             setter=self.set_setting,
                                             value=uid_info['properties']['value'])

        # The siren needs to be handled differently, as otherwise their binary async setting will not be named nicely
        if self.devices[device_uid].device_model_uid == "devolo.model.Siren":
            self.devices[device_uid].settings_property['muted'] = settings_property
        # As some devices have multiple binary async settings, we use the settings UID split after a '#' as key
        else:
            key = camel_case_to_snake_case(uid_info['UID'].split("#")[-1])
            self.devices[device_uid].settings_property[key] = settings_property

    def _last_activity(self, uid_info: dict):
        """
        Process last activity properties. Those don't go into an own property but will be appended to a parent property.
        This parent property is found by string replacement.
        """
        device_uid = get_device_uid_from_element_uid(uid_info['UID'])
        parent_element_uid = uid_info['UID'].replace("LastActivity", "BinarySensor")
        try:
            self.devices[device_uid].binary_sensor_property[parent_element_uid].last_activity = \
                uid_info['properties']['lastActivityTime']
        except AttributeError:
            parent_element_uid = uid_info['UID'].replace("LastActivity", "SirenMultiLevelSwitch")
            self.devices[device_uid].multi_level_switch_property[parent_element_uid].last_activity = \
                uid_info['properties']['lastActivityTime']

    def _led(self, uid_info: dict):
        """ Process LED information setting (lis) and visual feedback setting (vfs) properties. """
        device_uid = get_device_uid_from_setting_uid(uid_info['UID'])
        self._logger.debug("Adding led settings to %s.", device_uid)
        try:
            led_setting = uid_info['properties']['led']
        except KeyError:
            led_setting = uid_info['properties']['feedback']
        self.devices[device_uid].settings_property['led'] = SettingsProperty(element_uid=uid_info['UID'],
                                                                             setter=self.set_setting,
                                                                             led_setting=led_setting)

    def _meter(self, uid_info: dict):
        """ Process meter properties. """
        device_uid = get_device_uid_from_element_uid(uid_info['UID'])
        if not hasattr(self.devices[device_uid], "consumption_property"):
            self.devices[device_uid].consumption_property = {}
        self._logger.debug("Adding consumption property to %s.", device_uid)
        self.devices[device_uid].consumption_property[uid_info['UID']] = ConsumptionProperty(
            element_uid=uid_info['UID'],
            current=uid_info['properties']['currentValue'],
            total=uid_info['properties']['totalValue'],
            total_since=uid_info['properties']['sinceTime'])

    def _multilevel_async(self, uid_info: dict):
        """ Process multilevel async setting (mas) properties. """
        device_uid = get_device_uid_from_setting_uid(uid_info['UID'])
        try:
            name = camel_case_to_snake_case(uid_info['properties']['itemId'])
        # The Metering Plug has an multilevel async setting without an ID
        except TypeError:
            if self.devices[device_uid].device_model_uid == "devolo.model.Wall:Plug:Switch:and:Meter":
                name = "flash_mode"
            else:
                raise

        self._logger.debug("Adding %s setting to %s.", name, device_uid)
        self.devices[device_uid].settings_property[name] = SettingsProperty(element_uid=uid_info['UID'],
                                                                            setter=self.set_setting,
                                                                            value=uid_info['properties']['value'])

    def _multilevel_sync(self, uid_info: dict):
        """ Process multilevel sync setting (mss) properties. """
        device_uid = get_device_uid_from_setting_uid(uid_info['UID'])

        # The siren needs to be handled differently, as otherwise their multilevel sync setting will not be named nicely.
        if self.devices[device_uid].device_model_uid == "devolo.model.Siren":
            self._logger.debug("Adding tone settings to %s.", device_uid)
            self.devices[device_uid].settings_property['tone'] = SettingsProperty(element_uid=uid_info['UID'],
                                                                                  setter=self.set_setting,
                                                                                  tone=uid_info['properties']['value'])

        # The shutter needs to be handled differently, as otherwise their multilevel sync setting will not be named nicely.
        elif self.devices[device_uid].device_model_uid in ("devolo.model.OldShutter", "devolo.model.Shutter"):
            self._logger.debug("Adding shutter duration settings to %s.", device_uid)
            self.devices[device_uid].settings_property['shutter_duration'] = SettingsProperty(
                element_uid=uid_info['UID'],
                setter=self.set_setting,
                shutter_duration=uid_info['properties']['value'])

        # Other devices are up to now always motion sensors.
        else:
            self._logger.debug("Adding motion sensitivity settings to %s.", device_uid)
            self.devices[device_uid].settings_property['motion_sensitivity'] = SettingsProperty(
                element_uid=uid_info['UID'],
                setter=self.set_setting,
                motion_sensitivity=uid_info['properties']['value'])

    def _multi_level_sensor(self, uid_info: dict):
        """ Process multi level sensor properties. """
        device_uid = get_device_uid_from_element_uid(uid_info['UID'])
        if not hasattr(self.devices[device_uid], "multi_level_sensor_property"):
            self.devices[device_uid].multi_level_sensor_property = {}
        self._logger.debug("Adding multi level sensor property %s to %s.", uid_info['UID'], device_uid)
        self.devices[device_uid].multi_level_sensor_property[uid_info['UID']] = MultiLevelSensorProperty(
            element_uid=uid_info['UID'],
            value=uid_info['properties']['value'],
            unit=uid_info['properties']['unit'],
            sensor_type=uid_info['properties']['sensorType'])

    def _multi_level_switch(self, uid_info: dict):
        """ Process multi level switch properties. """
        device_uid = get_device_uid_from_element_uid(uid_info['UID'])
        if not hasattr(self.devices[device_uid], "multi_level_switch_property"):
            self.devices[device_uid].multi_level_switch_property = {}
        self._logger.debug("Adding multi level switch property %s to %s.", uid_info['UID'], device_uid)
        self.devices[device_uid].multi_level_switch_property[uid_info['UID']] = MultiLevelSwitchProperty(
            element_uid=uid_info['UID'],
            setter=self.set_multi_level_switch,
            value=uid_info['properties']['value'],
            switch_type=uid_info['properties']['switchType'],
            max=uid_info['properties']['max'],
            min=uid_info['properties']['min'])

    def _parameter(self, uid_info: dict):
        """ Process custom parameter setting (cps) properties."""
        device_uid = get_device_uid_from_setting_uid(uid_info['UID'])
        self._logger.debug("Adding parameter settings to %s.", device_uid)
        self.devices[device_uid].settings_property['param_changed'] = SettingsProperty(
            element_uid=uid_info['UID'],
            setter=self.set_setting,
            param_changed=uid_info['properties']['paramChanged'])

    def _protection(self, uid_info: dict):
        """ Process protection setting (ps) properties. """
        device_uid = get_device_uid_from_setting_uid(uid_info['UID'])
        self._logger.debug("Adding protection settings to %s.", device_uid)
        self.devices[device_uid].settings_property['protection'] = SettingsProperty(
            element_uid=uid_info['UID'],
            setter=self.set_setting,
            local_switching=uid_info['properties']['localSwitch'],
            remote_switching=uid_info['properties']['remoteSwitch'])

    def _remote_control(self, uid_info: dict):
        """ Process remote control properties. """
        device_uid = get_device_uid_from_element_uid(uid_info['UID'])
        self._logger.debug("Adding remote control to %s.", device_uid)
        if not hasattr(self.devices[device_uid], "remote_control_property"):
            self.devices[device_uid].remote_control_property = {}
        self.devices[device_uid].remote_control_property[uid_info['UID']] = RemoteControlProperty(
            element_uid=uid_info['UID'],
            setter=self.set_remote_control,
            key_count=uid_info['properties']['keyCount'],
            key_pressed=uid_info['properties']['keyPressed'],
            type=uid_info['properties']['type'])

    def _switch_type(self, uid_info: dict):
        """
        Process switch type setting (sts) properties. Interestingly, a switch with two buttons reports a switchType of 1 and a
        switch with four buttons reports a switchType of 2. This confusing behavior is corrected by doubling the value.
        """
        device_uid = get_device_uid_from_setting_uid(uid_info['UID'])
        self._logger.debug("Adding switch type setting to %s.", device_uid)
        self.devices[device_uid].settings_property['switch_type'] = SettingsProperty(element_uid=uid_info['UID'],
                                                                                     setter=self.set_setting,
                                                                                     value=uid_info['properties']['switchType']
                                                                                     * 2)

    def _temperature_report(self, uid_info: dict):
        """ Process temperature report setting (trs) properties. """
        device_uid = get_device_uid_from_setting_uid(uid_info['UID'])
        self._logger.debug("Adding temperature report settings to %s.", device_uid)
        self.devices[device_uid].settings_property['temperature_report'] = SettingsProperty(
            element_uid=uid_info['UID'],
            setter=self.set_setting,
            temp_report=uid_info['properties']['tempReport'],
            target_temp_report=uid_info['properties']['targetTempReport'])

    def _unknown(self, uid_info: dict):
        """ Ignore unknown properties. """
        ignore = ("devolo.SirenBinarySensor", "devolo.SirenMultiLevelSensor", "ss", "mcs")
        if not uid_info['UID'].startswith(ignore):
            self._logger.debug("Found an unexpected element uid: %s", uid_info['UID'])
