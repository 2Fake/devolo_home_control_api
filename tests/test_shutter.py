"""Test interacting with a shutter."""
import json
import sys

import pytest
from requests_mock import Mocker
from syrupy.assertion import SnapshotAssertion

from devolo_home_control_api.homecontrol import HomeControl

from . import HOMECONTROL_URL, Subscriber, load_fixture
from .mocks import WEBSOCKET

ELEMENT_ID = "hdm:ZWave:CBC56091/9"
FIXTURE = load_fixture("homecontrol_blinds")


@pytest.mark.skipif(sys.version_info < (3, 8), reason="Tests with snapshots need at least Python 3.8")
def test_getting_devices(local_gateway: HomeControl, snapshot: SnapshotAssertion) -> None:
    """Test getting shutter devices."""
    assert local_gateway.blinds_devices == snapshot


def test_state_change(local_gateway: HomeControl) -> None:
    """Test state change of a shutter."""
    subscriber = Subscriber(ELEMENT_ID)
    local_gateway.publisher.register(ELEMENT_ID, subscriber)
    shutter = local_gateway.devices[ELEMENT_ID]
    state = shutter.multi_level_switch_property[f"devolo.Blinds:{ELEMENT_ID}"]
    value = state.value
    last_activity = state.last_activity
    WEBSOCKET.recv_packet(json.dumps(FIXTURE["movement_event"]))
    assert value != state.value
    assert last_activity != state.last_activity
    assert subscriber.update.call_count == 1


def test_changing_general_device_settings(local_gateway: HomeControl) -> None:
    """Test changing general device settings."""
    shutter = local_gateway.devices[ELEMENT_ID]
    events_enabled = shutter.settings_property["general_device_settings"].events_enabled
    icon = shutter.settings_property["general_device_settings"].icon
    name = shutter.settings_property["general_device_settings"].name
    zone_id = shutter.settings_property["general_device_settings"].zone_id
    WEBSOCKET.recv_packet(json.dumps(FIXTURE["general_device_settings"]))
    assert events_enabled != shutter.settings_property["general_device_settings"].events_enabled
    assert icon != shutter.settings_property["general_device_settings"].icon
    assert name != shutter.settings_property["general_device_settings"].name
    assert zone_id != shutter.settings_property["general_device_settings"].zone_id


def test_setting_general_device_settings(remote_gateway: HomeControl, requests_mock: Mocker) -> None:
    """Test setting general device settings."""
    response = FIXTURE["success"]
    requests_mock.post(f"{HOMECONTROL_URL}/remote/json-rpc", json=response)
    shutter = remote_gateway.devices[ELEMENT_ID]
    events_enabled = shutter.settings_property["general_device_settings"].events_enabled
    icon = shutter.settings_property["general_device_settings"].icon
    name = shutter.settings_property["general_device_settings"].name
    zone_id = shutter.settings_property["general_device_settings"].zone_id
    shutter.settings_property["general_device_settings"].set(zone_id="hz_2", icon="icon_17")
    assert shutter.settings_property["general_device_settings"].events_enabled == events_enabled
    assert icon != shutter.settings_property["general_device_settings"].icon
    assert name == shutter.settings_property["general_device_settings"].name
    assert zone_id != shutter.settings_property["general_device_settings"].zone_id


def test_setting_automatic_calibration(remote_gateway: HomeControl) -> None:
    """Test setting automatic calibration settings."""
    shutter = remote_gateway.devices[ELEMENT_ID]
    automatic_calibration = shutter.settings_property["automatic_calibration"].calibration_status
    WEBSOCKET.recv_packet(json.dumps(FIXTURE["automatic_calibration"]))
    assert automatic_calibration != shutter.settings_property["automatic_calibration"].calibration_status
    WEBSOCKET.recv_packet(json.dumps(FIXTURE["automatic_calibration_old"]))
    assert automatic_calibration == shutter.settings_property["automatic_calibration"].calibration_status


def test_setting_movement_direction(remote_gateway: HomeControl) -> None:
    """Test setting the movement direction."""
    shutter = remote_gateway.devices[ELEMENT_ID]
    inverted = shutter.settings_property["movement_direction"].inverted
    WEBSOCKET.recv_packet(json.dumps(FIXTURE["movement_direction"]))
    assert inverted != shutter.settings_property["movement_direction"].inverted


def test_getting_properties(local_gateway: HomeControl) -> None:
    """Test getting a property of the device."""
    shutter = local_gateway.devices[ELEMENT_ID]
    assert shutter.get_property("multi_level_switch") == list(shutter.multi_level_switch_property.values())


def test_getting_state(local_gateway: HomeControl) -> None:
    """Test getting the online state of the device."""
    shutter = local_gateway.devices[ELEMENT_ID]
    assert shutter.is_online()
