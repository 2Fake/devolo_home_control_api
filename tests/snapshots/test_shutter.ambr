# serializer version: 1
# name: test_getting_devices
  list([
    Zwave(
      assistants_enabled=1,
      brand='Everspring',
      calibration_status=1,
      description=None,
      device_model_uid='devolo.model.Shutter',
      device_type='Door Lock Keypad',
      element_uids=list([
        'devolo.Blinds:hdm:ZWave:CBC56091/9',
      ]),
      generic_device_class='Entry Control',
      href='https://mydevolo.com/v1/zwave/products/0x0060/0x0001/0x0002',
      i2=True,
      identifier='SP814-US',
      inverted=False,
      is_own=True,
      is_securely_included=False,
      is_zwave_plus=True,
      manufacturer_id='0x0060',
      multi_level_switch_property=dict({
        'devolo.Blinds:hdm:ZWave:CBC56091/9': MultiLevelSwitchProperty(
          device_uid='hdm:ZWave:CBC56091/9',
          element_uid='devolo.Blinds:hdm:ZWave:CBC56091/9',
          last_activity=datetime.datetime(1970, 1, 1, 1, 0, tzinfo=tzfile('/usr/share/zoneinfo/Europe/Berlin')),
          max=100,
          min=0,
          switch_type='',
          unit='%',
          value=21,
        ),
      }),
      name='Everspring PIR Sensor SP814',
      pending_operations=False,
      product_id='0x0002',
      product_type_id='0x0001',
      secure_inclusion_code=0,
      setting_uids=list([
        'acs.hdm:ZWave:CBC56091/9',
        'bas.hdm:ZWave:CBC56091/9#i2',
        'bss.hdm:ZWave:CBC56091/9',
        'gds.hdm:ZWave:CBC56091/9',
        'mss.hdm:ZWave:CBC56091/9',
      ]),
      settings_property=dict({
        'automatic_calibration': SettingsProperty(
          calibration_status=True,
          element_uid='acs.hdm:ZWave:CBC56091/9',
          last_activity=datetime.datetime(1970, 1, 1, 1, 0, tzinfo=tzfile('/usr/share/zoneinfo/Europe/Berlin')),
        ),
        'general_device_settings': SettingsProperty(
          element_uid='gds.hdm:ZWave:CBC56091/9',
          events_enabled=True,
          icon='icon_16',
          last_activity=datetime.datetime(1970, 1, 1, 1, 0, tzinfo=tzfile('/usr/share/zoneinfo/Europe/Berlin')),
          name='Shutter',
          zone='Office',
          zone_id='hz_1',
        ),
        'i2': SettingsProperty(
          element_uid='bas.hdm:ZWave:CBC56091/9#i2',
          last_activity=datetime.datetime(1970, 1, 1, 1, 0, tzinfo=tzfile('/usr/share/zoneinfo/Europe/Berlin')),
          value=True,
        ),
        'movement_direction': SettingsProperty(
          element_uid='bss.hdm:ZWave:CBC56091/9',
          inverted=0,
          last_activity=datetime.datetime(1970, 1, 1, 1, 0, tzinfo=tzfile('/usr/share/zoneinfo/Europe/Berlin')),
        ),
        'shutter_duration': SettingsProperty(
          element_uid='mss.hdm:ZWave:CBC56091/9',
          last_activity=datetime.datetime(1970, 1, 1, 1, 0, tzinfo=tzfile('/usr/share/zoneinfo/Europe/Berlin')),
          shutter_duration=20,
        ),
      }),
      specific_device_class='Secure Keypad Door Lock',
      status=2,
      uid='hdm:ZWave:CBC56091/9',
      zwave_version='6.51.07',
    ),
  ])
# ---
